### PR TITLE
SWARM-242 adds Consul Ribbon support

### PR DESCRIPTION
**Pull request by jhovell**
@bobmcwhirter 

**What's here**
- Adds Consul support via  https://github.com/OrbitzWorldwide/consul-client (Apache license)
- Decouples Ribbon from JGroups/clustering by creating 2 different runtime modules for JGroups / Consul respectively
  - This meant changing RibbonFraction to RibbonJGroupsFraction - I realize this breaks examples, but I figured the naming was more intuitive, I can get a PR to update the examples put together if you think I am on the right track
- Uses the `org.jboss.msc.service` to instantiate the various singletons needed as with RibbonJGroups. This may be overkill, if some other approach is better, please advise.

**Limitations**
- Consul service advertising currently not supported, as I prefer registering services outside of the JVM where they can be healthchecked more universally, but am open to adding it. Advertising code moved to RibbonJGroupsFraction, as there is still some coupling to clustering anyway
- For the life of me, I absolutely could not get `javax.ws.rs.client.ClientBuilder.newBuilder()` to execute properly. It seemed to be that I could not write a module.xml to export the necessary extension in resteasy-client into the javax.ws.rs.api module. Typically I'd get some error such as `java.lang.ClassNotFoundException: org.glassfish.jersey.client.JerseyClientBuilder` which from reading `FactoryFinder.java` meant that the `javax.ws.rs.client.ClientBuilder` service extension could not be located. My workaround is to use `new RestEasyClientBuilder()` which is unideal. (See `ConsulClusterManagerActivator`)

This is my first time working with non-trivial JBoss modules, so to some degree I was finding my way in the dark. If you have suggestions on better ways to package this, possibly as a single fraction with different runtimes, let me know.

Also fixed is SSL support (`ribbon.IsSecure`). Several Ribbon module dependencies were missing.  I could add testing, but it requires setting up JKS certs & wasn't sure how to generate a cert within the context of a maven project. I could also just include a self-signed cert, but wasn't sure if that was overkill.

I could also add better Arquillian support, but I would need a package such as Wiremock to mock out having a Consul server present.

### CODE REVIEW

File: batch-jberet/test/pom.xml [hunk 1]
```diff
@@ -91,6 +91,11 @@
       <artifactId>weld-runtime</artifactId>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>org.wildfly.swarm</groupId>
+      <artifactId>remoting-runtime</artifactId>
```
Comment (jhovell) [inline batch-jberet/test/pom.xml:96]: unrelated, but seemed necessary for a successful build. Looks like currently builds are relying on having a previous maven SNAPSHOT artifact in the repo to pass

File: hawkular/test/pom.xml [hunk 2]
```diff
@@ -10,7 +10,7 @@
   <parent>
     <groupId>org.wildfly.swarm</groupId>
     <artifactId>hawkular-parent</artifactId>
-    <version>1.0.0.Alpha6-SNAPSHOT</version>
+    <version>1.0.0.Alpha7-SNAPSHOT</version>
```
Comment (jhovell) [inline hawkular/test/pom.xml:13]: unrelated, but was confusing my build, generating warnings

File: rxnetty/src/main/resources/modules/io/reactivex/rxnetty/main/module.xml [hunk 3]
```diff
@@ -10,6 +10,7 @@
     <module name="io.netty"/>
     <module name="com.netflix.servo"/>
     <module name="org.slf4j"/>
+    <module name="javax.api"/>
```
Comment (jhovell) [inline rxnetty/src/main/resources/modules/io/reactivex/rxnetty/main/module.xml:13]: necessary for SSL Ribbon

File: ribbon/modules/src/main/resources/modules/com/netflix/ribbon/main/module.xml [hunk 4]
```diff
@@ -6,6 +6,7 @@
     <artifact name="com.netflix.ribbon:ribbon-loadbalancer:${version.com.netflix.ribbon}"/>
     <artifact name="com.netflix.netflix-commons:netflix-commons-util:${version.com.netflix.commons}"/>
     <artifact name="com.netflix.netflix-commons:netflix-statistics:${version.com.netflix.commons}"/>
+    <artifact name="commons-lang:commons-lang:2.6"/>
```
Comment (jhovell) [inline ribbon/modules/src/main/resources/modules/com/netflix/ribbon/main/module.xml:9]: necessary for SSL Ribbon

**Comment by bobmcwhirter** [comment]
Fantastic. I'll try to review tomorrow. A PR to keep examples going would
be good.

## 

Sent from my iPhone

On Jan 20, 2016, at 4:54 PM, jhovell notifications@github.com wrote:

@bobmcwhirter https://github.com/bobmcwhirter

_What's here_
- Adds Consul support via
  https://github.com/OrbitzWorldwide/consul-client (Apache license)
- Decouples Ribbon from JGroups/clustering by creating 2 different
  runtime modules for JGroups / Consul respectively
  - This meant changing RibbonFraction to RibbonJGroupsFraction - I
    realize this breaks examples, but I figured the naming was more
    intuitive,
    I can get a PR to update the examples put together if you think
    I am on the
    right track
- Uses the org.jboss.msc.service to instantiate the various singletons
  needed as with RibbonJGroups. This may be overkill, if some other approach
  is better, please advise.

_Limitations_
- Consul service advertising currently not supported, as I prefer
  registering services outside of the JVM where they can be healthchecked
  more universally, but am open to adding it. Advertising code moved to
  RibbonJGroupsFraction, as there is still some coupling to clustering anyway
- For the life of me, I absolutely could not get
  javax.ws.rs.client.ClientBuilder.newBuilder() to execute properly. It
  seemed to be that I could not write a module.xml to export the necessary
  extension in resteasy-client into the javax.ws.rs.api module. Typically I'd
  get some error such as java.lang.ClassNotFoundException:
  org.glassfish.jersey.client.JerseyClientBuilder which from reading
  FactoryFinder.java meant that the javax.ws.rs.client.ClientBuilder
  service extension could not be located. My workaround is to use new
  RestEasyClientBuilder() which is unideal. (See
  ConsulClusterManagerActivator)

This is my first time working with non-trivial JBoss modules, so to some
degree I was finding my way in the dark. If you have suggestions on better
ways to package this, possibly as a single fraction with different
runtimes, let me know.

Also fixed is SSL support (ribbon.IsSecure). Several Ribbon module
dependencies were missing. I could add testing, but it requires setting up
JKS certs & wasn't sure how to generate a cert within the context of a
maven project. I could also just include a self-signed cert, but wasn't
sure if that was overkill.

I could also add better Arquillian support, but I would need a package such

## as Wiremock to mock out having a Consul server present.

You can view, comment on, or merge this pull request online at:

  https://github.com/wildfly-swarm/wildfly-swarm/pull/182
Commit Summary
- SWARM-242 adds Consul Ribbon support

File Changes
- _M_ batch-jberet/test/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-0
  (5)
- _M_ bom/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-1
  (10)
- _M_ fraction-list/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-2
  (7)
- _M_ hawkular/test/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-3
  (2)
- _M_ pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-4
  (23)
- _A_ ribbon-consul/api/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-5
  (43)
- _A_
  ribbon-consul/api/src/main/java/org/wildfly/swarm/netflix/ribbon/ConsulService.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-6
  (44)
- _A_
  ribbon-consul/api/src/main/java/org/wildfly/swarm/netflix/ribbon/RibbonConsulFraction.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-7
  (48)
- _A_ ribbon-consul/api/src/main/resources/wildfly-swarm-bootstrap.conf
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-8
  (1)
- _A_ ribbon-consul/modules/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-9
  (84)
- _A_
  ribbon-consul/modules/src/main/resources/modules/com/hashicorp/consul/main/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-10
  (20)
- _A_
  ribbon-consul/modules/src/main/resources/modules/org/wildfly/swarm/netflix/ribbon/consul/api/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-11
  (12)
- _A_
  ribbon-consul/modules/src/main/resources/modules/org/wildfly/swarm/netflix/ribbon/consul/main/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-12
  (15)
- _A_
  ribbon-consul/modules/src/main/resources/modules/org/wildfly/swarm/netflix/ribbon/consul/runtime/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-13
  (32)
- _A_ ribbon-consul/modules/src/main/resources/provided-dependencies.txt
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-14
  (7)
- _A_ ribbon-consul/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-15
  (32)
- _A_ ribbon-consul/runtime/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-16
  (77)
- _A_
  ribbon-consul/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/ConsulClusterManagerActivator.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-17
  (77)
- _A_
  ribbon-consul/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/RibbonConsulCacheListener.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-18
  (57)
- _A_
  ribbon-consul/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/RibbonConsulConfiguration.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-19
  (65)
- _R_ ribbon-consul/test/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-20
  (12)
- _R_
  ribbon-consul/test/src/test/java/org/wildfly/swarm/ribbon/RibbonConsulArquillianTest.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-21
  (10)
- _R_
  ribbon-consul/test/src/test/java/org/wildfly/swarm/ribbon/RibbonConsulInVmTest.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-22
  (8)
- _A_ ribbon-jgroups/api/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-23
  (73)
- _R_
  ribbon-jgroups/api/src/main/java/org/wildfly/swarm/netflix/ribbon/RibbonJGroupsArchive.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-24
  (8)
- _R_
  ribbon-jgroups/api/src/main/java/org/wildfly/swarm/netflix/ribbon/RibbonJGroupsArchiveImpl.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-25
  (12)
- _R_
  ribbon-jgroups/api/src/main/java/org/wildfly/swarm/netflix/ribbon/RibbonJGroupsConfigAsset.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-26
  (4)
- _R_
  ribbon-jgroups/api/src/main/java/org/wildfly/swarm/netflix/ribbon/RibbonJGroupsFraction.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-27
  (2)
- _A_
  ribbon-jgroups/api/src/main/resources/META-INF/services/org.wildfly.swarm.netflix.ribbon.RibbonJGroupsArchive
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-28
  (3)
- _A_ ribbon-jgroups/api/src/main/resources/wildfly-swarm-bootstrap.conf
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-29
  (1)
- _R_
  ribbon-jgroups/api/src/test/java/org/wildfly/swarm/netflix/ribbon/RibbonJGroupsArchiveTest.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-30
  (24)
- _A_ ribbon-jgroups/modules/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-31
  (100)
- _A_
  ribbon-jgroups/modules/src/main/resources/modules/org/wildfly/swarm/netflix/ribbon/jgroups/api/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-32
  (12)
- _A_
  ribbon-jgroups/modules/src/main/resources/modules/org/wildfly/swarm/netflix/ribbon/jgroups/main/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-33
  (15)
- _A_
  ribbon-jgroups/modules/src/main/resources/modules/org/wildfly/swarm/netflix/ribbon/jgroups/runtime/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-34
  (19)
- _A_ ribbon-jgroups/modules/src/main/resources/provided-dependencies.txt
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-35
  (7)
- _A_ ribbon-jgroups/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-36
  (32)
- _A_ ribbon-jgroups/runtime/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-37
  (71)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/AdvertiseCommand.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-38
  (1)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/ApplicationAdvertiser.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-39
  (0)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/ApplicationAdvertiserActivator.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-40
  (4)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/ClusterManager.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-41
  (0)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/ClusterManagerActivator.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-42
  (0)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/RequestAdvertisementsCommand.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-43
  (0)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/RibbonJGroupsConfiguration.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-44
  (23)
- _R_
  ribbon-jgroups/runtime/src/main/java/org/wildfly/swarm/netflix/ribbon/runtime/UnadvertiseCommand.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-45
  (0)
- _A_ ribbon-jgroups/test/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-46
  (62)
- _A_
  ribbon-jgroups/test/src/test/java/org/wildfly/swarm/ribbon/RibbonJGroupsArquillianTest.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-47
  (54)
- _A_
  ribbon-jgroups/test/src/test/java/org/wildfly/swarm/ribbon/RibbonJGroupsInVmTest.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-48
  (46)
- _M_ ribbon-webapp/api/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-49
  (2)
- _M_ ribbon-webapp/modules/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-50
  (2)
- _M_
  ribbon-webapp/modules/src/main/resources/modules/org/wildfly/swarm/ribbon/webapp/runtime/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-51
  (3)
- _M_ ribbon-webapp/runtime/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-52
  (2)
- _M_
  ribbon-webapp/runtime/src/main/java/org/wildfly/swarm/ribbon/webapp/runtime/RibbonWebAppConfiguration.java
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-53
  (5)
- _M_ ribbon/api/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-54
  (4)
- _D_
  ribbon/api/src/main/resources/META-INF/services/org.wildfly.swarm.netflix.ribbon.RibbonArchive
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-55
  (3)
- _M_ ribbon/modules/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-56
  (4)
- _M_
  ribbon/modules/src/main/resources/modules/com/netflix/ribbon/main/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-57
  (1)
- _M_ ribbon/pom.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-58
  (1)
- _M_
  rxnetty/src/main/resources/modules/io/reactivex/rxnetty/main/module.xml
  https://github.com/wildfly-swarm/wildfly-swarm/pull/182/files#diff-59
  (1)

Patch Links:
- https://github.com/wildfly-swarm/wildfly-swarm/pull/182.patch
- https://github.com/wildfly-swarm/wildfly-swarm/pull/182.diff

—
Reply to this email directly or view it on GitHub
https://github.com/wildfly-swarm/wildfly-swarm/pull/182.

**Comment by bobmcwhirter** [comment]
Heya--

I'm going to merge this locally and then see if we can't split apart Ribbon to be a single thing, and have a variety of Discovery fractions to supply the ServerLists.

So, a use would be able to use our :ribbon artifact, and maybe by default it assumes Eureka.

If they also include :discovery-jgroups, then our current jgroups-based discovery would be usable.

If they include :discovery-consul, then they get consul.

But I've not looked too deeply yet.  Thanks for the contribution!

**Comment by jhovell** [comment]
@bobmcwhirter that sounds like a great plan.

On a related note, have you thought about configuration support? Given there is already Archaius in the project, what do you think of building in support for a configuration server? Again, I am thinking Consul, but support for Zookeeper, maybe even something simple like Github might make sense as a key/value store. Also secret support? Apache Deltaspike has some interesting CDI configuration support, thought there might be a nice way to tie it all together in a CDI-aware JEE friendly way.

I'll work on adding a wildfly-swarm-examples PR, but I'll be out of the office tomorrow / off the grid until Sunday, hopefully it gets done this afternoon, if not by Monday for sure.

**Comment by bobmcwhirter** [comment]
So I dove in, and I'm trying to create a more bonafide SPI for discovery/topology management.  I'll scrape your Consul work over to it.

I'll also take responsibility for keeping the -examples happy when I push this, so you don't need to worry about that.

I am 100% clueless regarding configuration wrt Zookeeper, and only know Archaius from Ribbon's own requirements of it.  Probably good to open an issue at our JIRA to track thoughts around that.

**Comment by jhovell** [comment]
ok thanks @bobmcwhirter I'll save my time then. 

I created this issue to track dynamic property management. I'll post in the Google groups as well & maybe we can continue the discussion there.

https://issues.jboss.org/browse/SWARM-263

**Comment by bobmcwhirter** [comment]
So, here's what we've got at the moment:

https://github.com/wildfly-swarm/wildfly-swarm/commit/5cf64db85

Which has completely separated Ribbon bits from Topology bits.

I've knocked the JGroups implementation over to this:

https://github.com/wildfly-swarm/wildfly-swarm/blob/master/topology-jgroups/runtime/src/main/java/org/wildfly/swarm/topology/jgroups/runtime/JGroupsTopologyConnector.java

Basically we just need to implement a Consul (and Eurkeka) TopologyConnector:

https://github.com/wildfly-swarm/wildfly-swarm/blob/master/topology/runtime/src/main/java/org/wildfly/swarm/topology/runtime/TopologyConnector.java

With enough fraction/configuration/service-activator to stand it up under the well-known SERVICE_NAME.

Such as:

https://github.com/wildfly-swarm/wildfly-swarm/tree/master/topology-jgroups/runtime/src/main/java/org/wildfly/swarm/topology/jgroups/runtime

Next week I plan to move your Consul work to this SPI, and see if any further adjustments are necessary.

The upside of this is that we do not have a proliferation of RibbonWhatever fractions. There's still just Ribbon.

A TopologyConnector can certainly treat advertise() as a no-op, but if it _can_ advertise, I think it'd be good to implement that portion.  Else, it's a read-only topology, really. (or externally-managed).

In general, I think these changes should make it easier to slot in other topology-management mechanisms.

One major change, which might break your apps, is `ribbon-webapp` has become `topology-webapp` and the `/ribbon/ribbon.js` has become `/topology/topology.js`.

Another benefit though, is that topology bits can be used _without_ Ribbon at all, if there's any need for that, or if someone wants to use a different load-balancing client other than Ribbon.

**Comment by jhovell** [comment]
@bobmcwhirter this looks like a great improvement... definitely closer to what I had in mind. Thanks for all the work!

As for TopologyConnector / service advertising... With Eureka or Consul advertising involves a TTL such that services that no longer provide a heartbeat check in will be automatically unadvertised/deregistered from the topology service... it might be a really good idea for TopologyConnector to understand this, perhaps even with a callback for healthcheck status.

https://github.com/Netflix/eureka/wiki/Understanding-eureka-client-server-communication#renew 

Maybe this isn't relevant in the jgroups example, but I think it is fairly critical if registration is to be done in-app.

Also I digress as I don't know if the Swarm project has plans for adding some built-in healthchecks (a nice feature of Spring Cloud) but if so this might tie in neatly with deregistering if any healthcheck fails.

What do you want me to do with this PR? Do you want me to rebase your commit into it? Should I close it and you open another? 

Also, I can certainly re-add my Consul work on top of yours if you're short on time.

Minor, but in the meantime it might be useful to correct some of the Ribbon SSL module dependency issues as well.

**Comment by bobmcwhirter** [comment]
I’m currently working on copy/pasta to get your Consul stuff in.  You can
leave this PR open until I commit enough to satisfy it.

A separate PR for Ribbon SSL stuff would be good.

And a JIRA for health-checks in general would be useful.

I’ll work on adding a TTL/healthcheck to the topology SPI bits as I go with
your Consul stuff.

**Comment by bobmcwhirter** [comment]
I think this has been accomplished via:

https://github.com/wildfly-swarm/wildfly-swarm/commit/96e1947b66301d5a9ebed0b1ec522dc0cb3ef424

Closing this PR. Let's do another for the SSL bits and any improvements to the Consul support in general.

Thanks again!

**Pull request closed without merging**